### PR TITLE
new JSON formats for some DAML-LF data

### LINKED ISSUE #2361: Enhance Navigator's `lf-value-json` encoding to fully support DAML LF primitives
Large `Decimal` values is one of the cases.
See #2519

### PR DESCRIPTION
**Pull request by S11001001**
The `ApiCodecCompressed` uses some formats by default we don't want to use anymore, like `{"Some": {}}` for `Some ()`. We'll make the encoder use new formats, keeping the branches for the previous formats whenever no inconsistency arises, but adding new branches.

* [x] int64s from JsNumber
* [x] int64 to Number encode mode
* [x] decimals from JsNumber
* [x] decimal to Number encode mode
* [x] nullable root Optional
* [x] nested Optional as lists
* [x] decimal banker's rounding
* [x] reject non-Z timestamps
* [x] absent record fields in object syntax are `None`

Fixes #2361. Documented in #2519.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger-service/lf-value-json/src/main/scala/com/digitalasset/daml/lf/value/json/ApiCodecCompressed.scala [hunk 1]
```diff
@@ -97,96 +108,130 @@ abstract class ApiCodecCompressed[Cid](
...
-      case (JsArray(v), Model.DamlLfPrimType.List) =>
-        V.ValueList(v.map(e => jsValueToApiValue(e, prim.typArgs.head, defs)).to[FrontStack])
-      case (JsObject(f), Model.DamlLfPrimType.Optional) =>
-        f.headOption match {
-          case Some((`fieldNone`, _)) => V.ValueOptional(None)
-          case Some((`fieldSome`, v)) =>
-            V.ValueOptional(Some(jsValueToApiValue(v, prim.typArgs.head, defs)))
-          case Some(_) => deserializationError(s"Can't read ${value.prettyPrint} as Optional")
-          case None => deserializationError(s"Can't read ${value.prettyPrint} as Optional")
+    (prim.typ, value).match2 {
```
Comment (rautenrieth-da) [inline ledger-service/lf-value-json/src/main/scala/com/digitalasset/daml/lf/value/json/ApiCodecCompressed.scala]: Using `match2` is to preserve exhaustiveness checking? Or is there any other advantage?

File: ledger-service/lf-value-json/src/main/scala/com/digitalasset/daml/lf/value/json/ApiCodecCompressed.scala [hunk 2]
```diff
@@ -198,9 +243,7 @@ abstract class ApiCodecCompressed[Cid](
       case prim: Model.DamlLfTypePrim =>
         jsValueToApiPrimitive(value, prim, defs)
       case typeCon: Model.DamlLfTypeCon =>
-        val id = Model.DamlLfIdentifier(
-          typeCon.name.identifier.packageId,
-          typeCon.name.identifier.qualifiedName)
+        val id = typeCon.name.identifier
         // val dt = typeCon.instantiate(defs(id).getOrElse(deserializationError(s"Type $id not found")))
         val dt = Model.damlLfInstantiate(
```
Comment (rautenrieth-da) [inline ledger-service/lf-value-json/src/main/scala/com/digitalasset/daml/lf/value/json/ApiCodecCompressed.scala]: Unrelated to this commit: AFAIR `Model.damlLfInstantiate` is a copy of `typeCon.instantiate`, introduced when the daml.lf version had bug and Navigator could not upgrade to the latest daml.lf library version.

New issue for fixing this: https://github.com/digital-asset/daml/issues/2506

File: ledger-service/lf-value-json/src/main/scala/com/digitalasset/daml/lf/value/json/ApiCodecCompressed.scala (hunk 1 above)
Comment (S11001001) [reply ledger-service/lf-value-json/src/main/scala/com/digitalasset/daml/lf/value/json/ApiCodecCompressed.scala]: Yes; additionally, we might decide to resolve the first match layer in advance of having the `value` in the future, notwithstanding DAML recursive cases.

**Comment by S11001001** [comment]
> What are the use cases for enabling `encodeDecimalAsString` and `encodeInt64AsString`? Storing data in a database?

@rautenrieth-da That's a good one. We do plan to replace extractor's JSON encoder with this one.

The real reason it appears in this PR is "that's what's written in the spec".

> Serializing numeric data as JsNumber is not safe unless you have total control over who is going to deserialize the JSON object (now or in the future).

I could equally say that a deserializer could discard every character in a number-string after the seventh. This is not without precedent 🙂 so I wouldn't say this is a reason it's "unsafe".

Instead, I would say that our choice of default for the HTTP JSON API (and Navigator) "avoids common mistakes", whereas alternate cases e.g. extractor will simply make alternate choices.

Even in the cases you're referring to, it's not necessarily a concern that some data loss might occur, because not all use cases require the JSON encoding of contracts to be injective. Consider an application that charts some decimal values from a set of contracts, for example. We might want to provide such applications a way to [explicitly] request number encoding in the future.

**Comment by S11001001** [comment]
@leo-da Please check b4c00727b4, too, which was added after your review.

**Comment by leo-da** [comment]
> Even in the cases you're referring to, it's not necessarily a concern that some data loss might occur, because not all use cases require the JSON encoding of contracts to be injective. Consider an application that charts some decimal values from a set of contracts, for example. We might want to provide such applications a way to [explicitly] request number encoding in the future.

@S11001001 I don't think we can make it injective, there always will be a case when
DAML Decimal 9007199254740991 represented as JSON String: "9007199254740991"
and the case when
DAML Text "9007199254740991" represented as the same JSON String "9007199254740991"
or do you mean we that there will be cases when we don't need to operate with large numbers?

**Comment by S11001001** [comment]
@leo-da I should have said "JSON encoding of contracts _of the same type_".

**Comment by leo-da** [comment]
In any case if it is up to me, I would keep the option to represent `DAML Decima` as `JsNumber`

**Pull request merged**